### PR TITLE
Update vec_int16_ppc.h to add merge even/odd and multiply high

### PR DESCRIPTION
**Pull request by munroesj52**
Plus POWER9 absolute difference, additional examples in the
\file doxygen text, and clean up doxygent comment alignment.
Add more unit tests to cover new functions.

	* src/vec_int16_ppc.h: Update \File Description.
	Add \section i16_endian_issues_0_0.
	Add \section int16_examples_0_1.
	Add \section int16_perf_0_0.
	Clean up doxygen comment alignment for consistency.
	For all function \brief descriptions add Latency and
	throughput tables.
	(vec_absduh, vec_mrgahh, vec_mrgalh, vec_mrgeh,
	vec_mrgoh, vec_mulhsh, vec_mulhuh, vec_muluhm,
	vec_vmrgeh, vec_vmrgoh): New Function.

	* src/Makefile.am (libvecdummy_la_SOURCES):
	Add testsuite/vec_int16_dummy.c
	* src/Makefile.in: Regenerate from automake.

	*src/testsuite/arith128_test_i16.c (test_muluhm, test_mulhsh,
	test_mulhuh, test_mrgeoh, test_mrgahlh): New functions.
	(test_vec_i16): Add calls to test_mrgeoh, test_mrgahlh,
	test_mulhuh, test_mulhsh, test_muluhm.

	*src/testsuite/vec_int16_dummy.c: New File.

Signed-off-by: Steven Munroe <munroesj52@gmail.com>

### CODE REVIEW

File: src/vec_int16_ppc.h [hunk 1]
```diff
@@ -61,9 +79,447 @@
...
+  // M= 52429, a=0, s=3
+  vui16_t magic = vec_splats ((unsigned short) 52429);
+  const int s = 3;
+
+  q = vec_mulhuh (magic, n);
+  return vec_srhi (q, s);
+}
+ * \endcode
+ *
+ * But we need also need the modulo to extract each digit. This simplest and oldest technique is to
```
Comment (ThinkOpenly) [inline src/vec_int16_ppc.h]: s/need also need/also need/

File: src/vec_int16_ppc.h [hunk 2]
```diff
@@ -61,9 +79,447 @@
...
+  return rem;
+}
+ * \endcode
+ *
+ * \note vec_sub() and vec_splats() are an existing
+ * altivec.h generic built-ins.
+ *
+ * \subsection int16_examples_0_1_2 Divide by constant 10000 example
+ * As we mentioned above, some divisors require an add before the shift
+ * as a correction. For example for 16-bit divide by 10000 the magic
```
Comment (ThinkOpenly) [inline src/vec_int16_ppc.h]: Sentence fragment "For example..." followed by "For the 16-bit unsigned..."  You might be able to delete the whole fragment.

File: src/vec_int16_ppc.h [hunk 3]
```diff
@@ -61,9 +79,447 @@
...
+ * So the compiler schedules them to issue in same cycle.
+ * The latency for vmulouh/vmuleuh is 7 cycles while the addis latency
+ * is only 2 cycles.  The dependent addi instruction can issue in
+ * the 3rd cycle, while vmulouh/vmuleuh are still executing.
+ * The addi also has a 2 cycle latency, so the dependent lvx can issue
+ * in the 5th cycle, while vmulouh/vmuleuh are still executing.
+ * The lvx has a latency of 5 cycles and will not complete execution
+ * until 2 cycles after vmulouh/vmuleuh.
+ * The dependent xxlnor is waiting of the load (lvx)
+ * and has a latency of 2 cycles.
```
Comment (ThinkOpenly) [inline src/vec_int16_ppc.h]: You're enjoying this, aren't you?  ;-)

File: src/vec_int16_ppc.h [hunk 4]
```diff
@@ -148,6 +609,197 @@ vec_clzh (vui16_t vra)
...
+#if __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__
+  return vec_mrgahh (vec_mulo (vra, vrb), vec_mule (vra, vrb));
+#else
+  return vec_mrgahh (vec_mule (vra, vrb), vec_mulo (vra, vrb));
+#endif
+}
+
+/** \brief Vector Multiply Unsigned halfword Modulo.
+ *
+ *  Multiple the corresponding halfword elements of two vector unsigned
```
Comment (ThinkOpenly) [inline src/vec_int16_ppc.h]: s/Multiple/Multiply/

File: src/vec_int16_ppc.h [hunk 5]
```diff
@@ -387,4 +1064,88 @@ vec_srahi (vi16_t vra, const unsigned int shb)
...
+ *
+ * @param vra 128-bit vector unsigned short.
+ * @param vrb 128-bit vector unsigned short.
+ * @return A vector merge from only the odd halfwords of vra and vrb.
+ */
+static inline vui16_t
+vec_vmrgoh (vui16_t vra, vui16_t vrb)
+{
+#if __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__
+  const vui16_t permute =
```
Comment (ThinkOpenly) [inline src/vec_int16_ppc.h]: Here, like well above, it seems that declaring and initializing `permute` as `vui8_t` would be easier to read. (4 lines below as well.)

File: src/testsuite/arith128_test_i16.c [hunk 6]
```diff
@@ -209,6 +209,224 @@ test_revbh (void)
...
+  j = (vui16_t) {10, 20, 30, 40, 50, 60, 70, 80};
+  e = (vui16_t) {10, 40, 90, 160, 250, 360, 490, 640};
+  k = vec_muluhm (i, j);
+
+#ifdef __DEBUG_PRINT__
+  print_vint16d ("muluhm(\t{1, 2, 3, 4, ....},\n\t {10, 20, 30, 40, ...})\n\t ", (vui16_t)k);
+#endif
+  rc += check_vuint128x ("vec_muluhm:", (vui128_t)k, (vui128_t) e);
+
+  i = (vui16_t ){-1, -2, -3, -4, -5, -6, -7, -8};
```
[the remark this answers is not in the document]
Comment (munroesj52) [reply src/testsuite/arith128_test_i16.c]: Yes it is well defined in C and GCC (6.59.22 PowerPC AltiVec Built-in Functions)

Here we are testing a multiply low or multiply unsigned halfword module which do the same thing. The PowerISA always includes a statement of the form:

<multiply modulo> can be used for unsigned or signed integers.

No change planned

File: src/testsuite/arith128_test_i16.c [hunk 7]
```diff
@@ -209,6 +209,224 @@ test_revbh (void)
...
+  i = (vi16_t) CONST_VINT128_H(-256, -256, -256, -256, -256, -256, -256, -256);
+  j = (vi16_t) CONST_VINT128_H(-256, -512, -768, -1024, -1280, -1536, -1792, -2048);
+  e = (vi16_t) CONST_VINT128_H(1, 2, 3, 4, 5, 6, 7, 8);
+  k = vec_mulhsh (i, j);
+
+#ifdef __DEBUG_PRINT__
+  print_vint16d ("mulhsh(\t{-256,-256, ...},\n\t {-2^8,-2^9,-3*2^8,-2^10, ...})\n\t ", (vui16_t)k);
+#endif
+  rc += check_vuint128x ("vec_mulhsh:", (vui128_t)k, (vui128_t) e);
+
```
[the remark this answers is not in the document]
Comment (munroesj52) [reply src/testsuite/arith128_test_i16.c]: Ok will this do:

  i = (vi16_t) CONST_VINT128_H(-32768,-32768,-32768,-32768, 32767, 32767, 32767, 32767);
  j = (vi16_t) CONST_VINT128_H(-32768, 32767,-16384, 16384,-32768, 32767,-16384, 16384);
  e = (vi16_t) CONST_VINT128_H( 16384,-16384,  8192, -8192,-16384, 16383, -8192,  8191);

File: src/testsuite/arith128_test_i16.c [hunk 8]
```diff
@@ -209,6 +209,224 @@ test_revbh (void)
...
+  i = (vui16_t) CONST_VINT128_H(10000, 10000, 10000, 10000, 10000, 10000, 10000, 10000);
+  j = (vui16_t) CONST_VINT128_H(7, 68, 661, 6554, 7, 68, 661, 6554);
+  e = (vui16_t) CONST_VINT128_H(1, 10, 100, 1000, 1, 10, 100, 1000);
+  k = vec_mulhuh (i, j);
+
+#ifdef __DEBUG_PRINT__
+  print_vint16d ("mulhuh(\t{10^4,10^4,10^4,10^4, ...},\n\t {10^-4,10^-3,10^-2,10^-1})\n\t ", k);
+#endif
+  rc += check_vuint128x ("vec_mulhuh:", (vui128_t)k, (vui128_t) e);
+
```
[the remark this answers is not in the document]
Comment (munroesj52) [reply src/testsuite/arith128_test_i16.c]: Ok how about:

  i = (vui16_t) CONST_VINT128_H(65535, 65535, 65535, 65535, 32768, 32767, 16384, 16383);
  j = (vui16_t) CONST_VINT128_H(65535, 32768, 16384, 16383, 32768, 16384, 16384, 16383);
  e = (vui16_t) CONST_VINT128_H(65534, 32767, 16383, 16382, 16384,  8191,  4096,  4095);

File: src/vec_int16_ppc.h [hunk 9]
```diff
@@ -61,9 +79,447 @@
...
+ * results in LE mode.
+ *
+ * The first step is the provide implement Vector Merge Even Halfword operation:
+ * \code
+static inline vui16_t
+vec_vmrgeh (vui16_t vra, vui16_t vrb)
+{
+#if __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__
+  vui16_t permute =
+      { 0x0302,0x1312, 0x0706,0x1716, 0x0B0A,0x1B1A, 0x0F0E,0x1F1E };
```
[the remark this answers is not in the document]
Comment (munroesj52) [reply src/vec_int16_ppc.h]: vec_perm requires that the permute select vector to be type vector char. This is independent of the type of the input vectors. In this case the input vector are type vector unsigned halfword (vui16_t).

This is hard enough with all the endian noise. So I use mental tricks to help understand what I am trying to do. Organizing the permute select vector by halfwords (to match the input) helps ME.

No plan to change this.

File: src/vec_int16_ppc.h [hunk 10]
```diff
@@ -61,9 +79,447 @@
...
+
+  return vec_perm (vrb, vra, (vui8_t)permute);
+#else
+  vui16_t permute =
+      { 0x0001,0x1011, 0x0405,0x1415, 0x0809,0x1819, 0x0C0D,0x1C1D};
+
```
[the remark this answers is not in the document]
Comment (munroesj52) [reply src/vec_int16_ppc.h]: not really, the short constants are in magnitude order not element order. So your initializer would give wrong results. See previous comment.

No plan to change this.

File: src/vec_int16_ppc.h [hunk 11]
```diff
@@ -61,9 +79,447 @@
...
+ *
+ * As good OpenPOWER ABI citizens we should also provide endian
+ * sensitive operations vec_mrgeh() vec_mrgoh().
+ * For example:
+ * \code
+static inline vui16_t
+vec_mrgeh  (vui16_t vra, vui16_t vrb)
+{
+#if __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__
+  return vec_vmrgoh ((vui16_t) vrb, (vui16_t) vra);
```
[the remark this answers is not in the document]
Comment (munroesj52) [reply src/vec_int16_ppc.h]: Ok

File: src/vec_int16_ppc.h [hunk 12]
```diff
@@ -61,9 +79,447 @@
...
+{
+  return vec_vmrgeh ((vui16_t) vra, (vui16_t) vrb);
+}
+ * \endcode
+ * This is simpler as we can use the endian invariant vec_vmrgeh() operation.
+ * Similarly for Vector Merge Algebraic Low Halfword using vec_vmrgoh().
+ * \note The inputs are defined as 32-bit to match the results from
+ * multiply even/odd halfword.
+ *
+ * How we have all the parts we need to implement multiply high/low halfword.
```
[the remark this answers is not in the document]
Comment (munroesj52) [reply src/vec_int16_ppc.h]: OK

File: src/vec_int16_ppc.h [hunk 13]
```diff
@@ -61,9 +79,447 @@
...
+ * \endcode
+ * The vmulouh, vmuleuh, vperm instruction sequence is the core of the
+ * function.  They multiply the elements and selects/merges the high
+ * order 16-bits of each product into the result vector.
+ *
+ * \section int16_examples_0_1 Examples, Divide by integer constant
+ * Suppose we have a requirement to convert an array of 16-bit
+ * unsigned short values to decimal. The classic <I>itoa</I>
+ * implementation performs a sequence of divide / modulo by 10
+ * operations that produce on (decimal) per iteration,
```
[the remark this answers is not in the document]
Comment (munroesj52) [reply src/vec_int16_ppc.h]: OK

File: src/vec_int16_ppc.h [hunk 14]
```diff
@@ -61,9 +79,447 @@
...
+ * until the divide returns 0.
+ *
+ * For this example we want to vectorize the operation and the
+ * PowerISA (and and most other platforms) does not provide a
```
[the remark this answers is not in the document]
Comment (munroesj52) [reply src/vec_int16_ppc.h]: OK

File: src/vec_int16_ppc.h [hunk 15]
```diff
@@ -61,9 +79,447 @@
...
+ * vector integer divide instruction. But we do have vector integer
+ * multiply as we will see the multiply high defined above is very
```
[the remark this answers is not in the document]
Comment (munroesj52) [reply src/vec_int16_ppc.h]: OK

File: src/vec_int16_ppc.h [hunk 16]
```diff
@@ -61,9 +79,447 @@
...
+ * handy for applying the multiplicative inverse.
+ * Also we the conversion divide is a constant value applied across
```
[the remark this answers is not in the document]
Comment (munroesj52) [reply src/vec_int16_ppc.h]: OK

File: src/vec_int16_ppc.h [hunk 17]
```diff
@@ -61,9 +79,447 @@
...
+ * the 16-bit fraction and we have a multiply high (vec_mulhuh())
+ * operation. Multiplying a 16-bit unsigned integer by a 16-bit
+ * unsigned fraction generates a 32-bit product with 16-bits above
+ * (integer) and below (fraction) the radix point. The high 16-bits
+ * of the product is a good approximation of the integer quotient.
+ *
+ * It turns out that generating the multiplicative inverse can be
+ * tricky.  To produce correct results over the full range, requires
+ * possible pre-scaling and post-shifting, and sometimes a corrective
+ * addition. Fortunately the mathematics are well
```
[the remark this answers is not in the document]
Comment (munroesj52) [reply src/vec_int16_ppc.h]: OK

File: src/vec_int16_ppc.h [hunk 18]
```diff
@@ -61,9 +79,447 @@
...
+ * \see "Hacker's Delight, 2nd Edition,"
+ * Henry S. Warren, Jr, Addison Wesley, 2013.
+ * Chapter 10, Integer Division by Constants.
+ *
+ * \subsection int16_examples_0_1_1 Divide by constant 10 examples
+ * In the chapter above;
+ * <BLOCKQUOTE>Figure 10-2 Computing the magic number for unsigned division.</BLOCKQUOTE>
+ * provides a sample C function for generating the magic number
+ * (actually a struct containing; the magic multiplicative inverse,
+ * "add" indicator, and the shift amount.).
```
[the remark this answers is not in the document]
Comment (munroesj52) [reply src/vec_int16_ppc.h]: OK

File: src/vec_int16_ppc.h [hunk 19]
```diff
@@ -61,9 +79,447 @@
...
+ * For the 16-bit unsigned divisor 10 this is {  52429, 0, 3 }:
```
[the remark this answers is not in the document]
Comment (munroesj52) [reply src/vec_int16_ppc.h]: OK

File: src/vec_int16_ppc.h (hunk 1 above)
Comment (munroesj52) [reply src/vec_int16_ppc.h]: Ok, also changed the s/This simplest/The simplest/

File: src/vec_int16_ppc.h (hunk 2 above)
Comment (munroesj52) [reply src/vec_int16_ppc.h]: OK

File: src/vec_int16_ppc.h (hunk 3 above)
Comment (munroesj52) [reply src/vec_int16_ppc.h]: No that there is anything wrong with that ;)

File: src/vec_int16_ppc.h (hunk 4 above)
Comment (munroesj52) [reply src/vec_int16_ppc.h]: Ok

File: src/vec_int16_ppc.h (hunk 5 above)
Comment (munroesj52) [reply src/vec_int16_ppc.h]: See reply for vec_vmrgeh above. This is hard enough and these are the tricks I use.

No plan to change this.

**Comment by munroesj52** [comment]
Soo ... I am trying to use (resolve conversation) format but I am not sure exactly who should reply when.

So:
- if I agree and I think it clear/obvious I replay OK and mark it resolved.
- If I disagree and explain why I did what I did and don't plan to change it now. I reply with the explanation and mark it resolved.
- If have a suggestion that I think will resolve you issue. I reply with the my alternative and leave it for you to mark it resolve or counter.

Does this make sense?

File: src/testsuite/arith128_test_i16.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (ThinkOpenly) [reply src/testsuite/arith128_test_i16.c]: yep!

File: src/testsuite/arith128_test_i16.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (ThinkOpenly) [reply src/testsuite/arith128_test_i16.c]: yep!

File: src/vec_int16_ppc.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (ThinkOpenly) [reply src/vec_int16_ppc.h]: ugh.  I see why you are doing that now.  Still pretty ugly, with magic numbers.  A suggestion for your consideration:
```
#define HINDEX_AS_BYTEINDICES(i) (((i * 2 + 1) << 8) + (i * 2))
vui16_t permute =
  {
    HINDEX_AS_BYTEINDICES(1),
    HINDEX_AS_BYTEINDICES(9),
    HINDEX_AS_BYTEINDICES(3),
    HINDEX_AS_BYTEINDICES(11),
    HINDEX_AS_BYTEINDICES(5),
    HINDEX_AS_BYTEINDICES(13),
    HINDEX_AS_BYTEINDICES(7),
    HINDEX_AS_BYTEINDICES(15)
  };
```

**Pull request merged**